node/modules/storageminer.go

```diff
@@ -357,76 +356,187 @@ func SectorStorage(mctx helpers.MetricsCtx, lc fx.Lifecycle, lstor *paths.Local,
 	return sst, nil
 }
 
-func F3Participation(mctx helpers.MetricsCtx, lc fx.Lifecycle, api v1api.FullNode, minerAddress dtypes.MinerAddress) error {
-	ctx := helpers.LifecycleCtx(mctx, lc)
-	b := &backoff.Backoff{
-		Min:    1 * time.Second,
-		Max:    1 * time.Minute,
-		Factor: 1.5,
-		Jitter: false,
+type f3Participator struct {
+	node                     v1api.FullNode
+	participant              address.Address
+	backoff                  *backoff.Backoff
+	timer                    *time.Timer
+	maxCheckProgressAttempts int
+	previousTicket           api.F3ParticipationTicket
+	checkProgressInterval    time.Duration
+	leaseTerm                uint64
+}
+
+func newF3Participator(node v1api.FullNode, participant dtypes.MinerAddress, backoff *backoff.Backoff, maxCheckProgress int, checkProgressInterval time.Duration, leaseTerm uint64) *f3Participator {
+	return &f3Participator{
+		node:                     node,
+		participant:              address.Address(participant),
+		backoff:                  backoff,
+		timer:                    time.NewTimer(0),
+		maxCheckProgressAttempts: maxCheckProgress,
+		checkProgressInterval:    checkProgressInterval,
+		leaseTerm:                leaseTerm,
 	}
-	go func() {
-		timer := time.NewTimer(0)
-		defer timer.Stop()
+}
 
-		if !timer.Stop() {
-			<-timer.C
+func (p *f3Participator) participate(ctx context.Context) error {
+	for ctx.Err() == nil {
+		if ticket, err := p.tryGetF3ParticipationTicket(ctx); err != nil {
+			return err
+		} else if lease, participating, err := p.tryF3Participate(ctx, ticket); err != nil {
+			return err
+		} else if !participating {
+			continue
+		} else if err := p.awaitLeaseExpiry(ctx, lease); err != nil {
+			return err
 		}
```

Comment:
Do we really want to exit this loop if we hit an error?

Reply:
Or are we saying that all errors are unrecoverable and that these APIs should only return errors when recovery is impossible?

Reply:
> Do we really want to exit this loop if we hit an error?

Yes because `awaitLeaseExpiry` exits with error only if either [F3 is disabled](https://github.com/filecoin-project/lotus/blob/3d12ca5c53c5df74bb9daa0ea9be27b283e2469c/node/modules/storageminer.go#L460) or [context has an error](https://github.com/filecoin-project/lotus/blob/3d12ca5c53c5df74bb9daa0ea9be27b283e2469c/node/modules/storageminer.go#L477).

> APIs should only return errors when recovery is impossible?

The APIs only return error when recovery is not possible, yes.